src/Workspaces/Core/MSBuild/MSBuild/MSBuildWorkspace_ProjectReload.cs

```diff
@@ -0,0 +1,309 @@
+                project = project.Solution.WithProjectAssemblyName(projectId, attributes.AssemblyName).GetProject(projectId)!;
+            }
+
+            if (project.State.ProjectInfo.Attributes.FilePath != attributes.FilePath)
+            {
+                project = project.Solution.WithProjectFilePath(projectId, attributes.FilePath).GetProject(projectId)!;
+            }
+
+            if (project.State.ProjectInfo.Attributes.OutputFilePath != attributes.OutputFilePath)
+            {
+                project = project.Solution.WithProjectOutputFilePath(projectId, attributes.OutputFilePath).GetProject(projectId)!;
+            }
+
+            if (project.State.ProjectInfo.Attributes.OutputRefFilePath != attributes.OutputRefFilePath)
+            {
+                project = project.Solution.WithProjectOutputRefFilePath(projectId, attributes.OutputRefFilePath).GetProject(projectId)!;
+            }
+
+            if (project.State.ProjectInfo.Attributes.CompilationOutputInfo != attributes.CompilationOutputInfo)
+            {
+                project = project.Solution.WithProjectCompilationOutputInfo(project.Id, attributes.CompilationOutputInfo).GetProject(project.Id)!;
+            }
+
+            if (project.State.ProjectInfo.Attributes.DefaultNamespace != attributes.DefaultNamespace)
+            {
+                project = project.Solution.WithProjectDefaultNamespace(projectId, attributes.DefaultNamespace).GetProject(projectId)!;
+            }
+
+            if (project.State.ProjectInfo.Attributes.HasAllInformation != attributes.HasAllInformation)
+            {
+                project = project.Solution.WithHasAllInformation(projectId, attributes.HasAllInformation).GetProject(projectId)!;
+            }
+
+            if (project.State.ProjectInfo.Attributes.RunAnalyzers != attributes.RunAnalyzers)
+            {
+                project = project.Solution.WithRunAnalyzers(projectId, attributes.RunAnalyzers).GetProject(projectId)!;
+            }
+
+            return project;
+        }
+
+        private static async Task<Project> UpdateDocumentsAsync(
+                Project project,
+                IReadOnlyList<DocumentInfo> documentInfos,
+                IEnumerable<TextDocumentState> existingTextDocumentStates,
+                ChecksumCollection oldChecksums,
+                ChecksumCollection newChecksums,
+                Func<Solution, ImmutableArray<DocumentInfo>, Solution> addDocuments,
+                Func<Solution, DocumentId, Solution> removeDocument)
+        {
+            using var olds = SharedPools.Default<HashSet<Checksum>>().GetPooledObject();
+            using var news = SharedPools.Default<HashSet<Checksum>>().GetPooledObject();
+
+            olds.Object.UnionWith(oldChecksums);
+            news.Object.UnionWith(newChecksums);
+
+            // remove documents that exist in both side
+            olds.Object.ExceptWith(newChecksums);
+            news.Object.ExceptWith(oldChecksums);
```

Comment:
What is ensuring consistent DocumentId generation? My gut would have been each time we're simply generating new IDs entirely, so the changing documents aren't being impacted?